### PR TITLE
Feat: Convert to ESLint Flat Config

### PR DESCRIPTION
**Pull request by bsokol-wl**
Rebuild the ESLint plugin to use flat config. Update documentation.

### CODE REVIEW

File: packages/stylelint-configs/eslint.config.mjs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import workleapPlugin from "@workleap/eslint-plugin";
```
Comment (alexasselin008) [inline packages/stylelint-configs/eslint.config.mjs:1]: FYI i think stylelint-configs can now be converted to an esm project since version 16

File: packages/eslint-plugin/lib/config/by-project-type/react-library.ts [hunk 2]
```diff
@@ -1,26 +1,31 @@
...
-        // Custom WorkLeap rules
-        "@workleap/strict-css-modules-names": "warn"
-    }
-};
+const config: Linter.FlatConfig[] = [
+    {
+        ignores: ["dist/"]
+    },
+    workleap,
+    ...core,
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/by-project-type/react-library.ts:19]: Is there a specific reason why we don't merge the globals for this one?

File: packages/eslint-plugin/lib/config/by-project-type/typescript-library.ts [hunk 3]
```diff
@@ -1,19 +1,24 @@
...
-        "plugin:@workleap/mdx",
-        "plugin:@workleap/package-json",
-        "plugin:@workleap/yaml"
-    ]
-};
+const config: Linter.FlatConfig[] = [
+    {
+        ignores: ["dist/"]
+    },
+    ...core,
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/by-project-type/typescript-library.ts:15]: Is there a specific reason why we don't merge the globals for this one?

File: packages/eslint-plugin/lib/config/by-project-type/monorepo-workspace.ts [hunk 4]
```diff
@@ -1,22 +1,30 @@
...
-    rules: {
-        "package-json/valid-version": "off"
-    }
-};
+const config: Linter.FlatConfig[] = [
+    {
+        ignores: ["dist/"]
+    },
+    ...(withGlobals(core, ["node"])),
+    ...(withGlobals(typescript, ["node"])),
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/by-project-type/monorepo-workspace.ts:16]: Why did you choose not to do the same as you did for the react config?

```
        languageOptions: {
            ...reactPlugin.configs.recommended.languageOptions,
            globals: {
                ...globals.serviceworker,
                ...globals.browser
            }
        },
```

File: packages/eslint-plugin/lib/config/storybook.ts [hunk 5]
```diff
@@ -1,30 +1,29 @@
...
+        plugins: { "storybook": storybookPlugin },
+        rules: {
+            ...storybookPlugin.configs.recommended.overrides[0].rules,
+            ...storybookPlugin.configs.csf.overrides[0].rules,
+            ...storybookPlugin.configs["csf-strict"].rules
+        }
+    },
+    {
+        files: mainStorybookFiles,
+        plugins: { "storybook": storybookPlugin },
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/storybook.ts:22]: Looking at the previous config, it seems like this section didn't have the storybook plugin configured?

File: packages/eslint-plugin/lib/config/testing-library.ts [hunk 6]
```diff
@@ -1,22 +1,22 @@
...
+import testingLibraryPlugin from "eslint-plugin-testing-library";
+import { reactTestFiles, testFiles } from "../utils/patterns.ts";
 
-const config: Linter.Config = {
-    overrides: [
-        {
-            files: reactTestFiles,
-            plugins: ["testing-library"],
-            extends: ["plugin:testing-library/react"]
+const config: Linter.FlatConfig[] = [
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/testing-library.ts:5]: I think it's supposed to use 2 different extends:

Here's the original config:

```
const config: Linter.Config = {
    overrides: [
        {
            files: reactTestFiles,
            plugins: ["testing-library"],
            extends: ["plugin:testing-library/react"]
        },
        {
            files: testFiles,
            extends: ["plugin:testing-library/dom"]
        }
    ]
};
```

File: packages/eslint-plugin/lib/config/typescript.ts [hunk 7]
```diff
@@ -1,76 +1,77 @@
...
-            ],
-            rules: {
-                // @typescript-eslint/recommended disables
-                "@typescript-eslint/no-non-null-assertion": "off",
+const config: Linter.FlatConfig[] = [
+    {
+        files: typeScriptFiles,
+        plugins: {
+            // @ts-ignore
+            "@typescript-eslint": tseslint.plugin
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/typescript.ts:10]: So I guess that because of this plugin, you don't need to extends `plugin:@typescript-eslint/eslint-recommended` and `plugin:@typescript-eslint/recommended` anymore?

By extend, I mean using the spread syntax in the `rules` section.

File: packages/eslint-plugin/lib/config/workleap.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import type { Linter } from "eslint";
+import StrictCSSModuleNames from "../rules/strict-css-modules-names.ts";
+
+const config: Linter.FlatConfig =
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/workleap.ts:4]: What are the drawbacks of importing the rule directly in the index file like it was done previously?

File: packages/eslint-plugin/lib/config/yaml.ts [hunk 9]
```diff
@@ -1,18 +1,11 @@
...
-            extends: ["plugin:yml/recommended"],
-            parser: "yaml-eslint-parser"
-        }
-    ]
-};
+const config: Linter.FlatConfig[] =
+    ymlPlugin.configs["flat/recommended"].map(conf => ({
+        ...conf,
+        files: yamlFiles
+    }));
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/yaml.ts:9]: Why are you using this "map" syntax instead of using a syntax similar to the other files?

File: packages/eslint-plugin/lib/utils/patterns.ts [hunk 10]
```diff
@@ -1,40 +1,45 @@
...
-    "*.ts?(x)"
+export const sourceFiles = [
+    ...javaScriptFiles,
+    ...typeScriptFiles
 ];
 
 export const testFiles = [
-    "*.test.[jt]s",
-    "*-test.[jt]s",
+    "**/*.test.[jt]s",
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/utils/patterns.ts:16]: You mentioned that you were going to revisit this, did you?

File: packages/stylelint-configs/eslint.config.mjs (hunk 1 above)
Comment (bsokol-wl) [reply packages/stylelint-configs/eslint.config.mjs:1]: I'd prefer to do that in a separate release if that's ok. Already a lot going on here.

File: docs/eslint/flat-config-migration.md [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+import workleapPlugin from "@workleap/eslint-config";
+
+const config = [
+
+];
+
+export default config;
+```
+
+You can choose to combine [any individual configs](wl-web-configs/eslint/advanced-composition/) or chose a one of the "by project type" configs (more info below). Here, we'll compose a config for a project that uses React and TypeScript. By convention, all configs are found at `workleapPlugin.configs`. Each config can be a single object or an array, but for simplicity, all Workleap configs are exported as arrays. Therefore, each Workleap config must be spread (`...`) into the config array.
```
[the remark this answers is not in the document]
Comment (bsokol-wl) [reply docs/eslint/flat-config-migration.md:37]: I'll rearrange it to focus on the project type

File: docs/eslint/flat-config-migration.md [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+Add 2 scripts to the top leve `package.json`:
+```json
+"scripts": {
+    "lint": "pnpm run \"/^lint:.*/\"",
+    "lint:eslint": "eslint . --max-warnings=0 --cache --cache-location node_modules/.cache/eslint",
+    "lint:eslint-packages": "pnpm -r --parallel --if-present --aggregate-output lint:eslint",
+}
+```
+- The `lint` script will run all scripts that start with "lint:" inside this `package.json`.
```
[the remark this answers is not in the document]
Comment (bsokol-wl) [reply docs/eslint/flat-config-migration.md:156]: I was basing that off the `lint:types` script which also has to run the script from each package.

The new "root" behavior is not exactly the same as the old "root" behavior for a config file, as I discovered. If you just run eslint in the root directory, it will use whatever config file is in the root directory, and ignore the ones in the monorepo packages.

So you have to have the root config ignore the packages, and then use pnpm to kick off eslint within each package. That's what's happening in parallel. In my tests, it's more than twice as fast as running eslint in each package sequentially.

[the remark this answers is not in the document]
Comment (patricklafrance) [reply docs/eslint/flat-config-migration.md:156]: Indeed its faster. The issue we were having with running eslint within each package is the way the errors were reported. It was outputed from PNPM CLI instead of ESLint CLI and the result was less good.

Regarding the root behavior, I am kind of surprised because we used to add "root" to all our nested ESLint configs.

File: packages/eslint-plugin/lib/utils/patterns.ts (hunk 10 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/utils/patterns.ts:16]: I did revisit this, they need the ** or they will only lint the top level files.

File: packages/eslint-plugin/lib/config/yaml.ts (hunk 9 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/yaml.ts:10]: Because the YAML config already has a flat config, and I just need to map over it and modify the files. It doesn't need any other updates.

File: packages/eslint-plugin/lib/config/workleap.ts (hunk 8 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/workleap.ts:4]: I believe I was avoiding a circular dependency by moving it to its own file

File: packages/eslint-plugin/lib/config/testing-library.ts (hunk 6 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/testing-library.ts:5]: I'll check, probably just missed the DOM one.

File: packages/eslint-plugin/lib/config/typescript.ts (hunk 7 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/typescript.ts:11]: I will need to check on this. They changed their structure but there is still a recommended config.

File: packages/eslint-plugin/lib/config/testing-library.ts (hunk 6 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/testing-library.ts]: The DOM rules are present in the second object.

File: packages/eslint-plugin/lib/config/typescript.ts (hunk 7 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/typescript.ts:16]: I've added the missing recommended rules

File: packages/eslint-plugin/lib/config/by-project-type/monorepo-workspace.ts [hunk 13]
```diff
@@ -1,22 +1,30 @@
...
-        "plugin:@workleap/package-json",
-        "plugin:@workleap/yaml"
-    ],
-    rules: {
-        "package-json/valid-version": "off"
-    }
-};
+const config: Linter.FlatConfig[] = [
+    {
+        ignores: ["dist/"]
```
[the remark this answers is not in the document]
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/by-project-type/monorepo-workspace.ts]: Forgot to respond to this before but I added one to the root

File: packages/eslint-plugin/lib/config/storybook.ts (hunk 5 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/storybook.ts:22]: All this is doing is providing the rule definitions. I'm not clear on if this is necessary since it's also provided above. I believe it is, since it's for different file patterns. But even if it isn't necessary, it isn't hurting anything being there.

File: packages/eslint-plugin/lib/config/core.ts [hunk 14]
```diff
@@ -1,158 +1,144 @@
...
-                node: true,
-                commonjs: true
-            },
-            rules: {
-                // eslint:recommended overwrite some rules
-                "no-cond-assign": ["error", "except-parens"],
-                "no-labels": ["warn", { allowLoop: true, allowSwitch: false }],
-                "no-prototype-builtins": "off",
+const config: Linter.FlatConfig[] = await concat(
+    js.configs.recommended as Linter.FlatConfig,
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/core.ts:8]: Is the "as Linter.FlatConfig" still required?

File: packages/eslint-plugin/lib/config/core.ts [hunk 15]
```diff
@@ -1,158 +1,144 @@
...
+    {
+        name: "Workleap/Core",
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/core.ts:10]: Is "Workleap/Core" matching a recommended format pushed by eslint?

If not, could you update to lowercase and use "-" between words please? 🙏🏻 Ex.: "workleap-core"

Comment (patricklafrance) [reply packages/eslint-plugin/lib/config/core.ts:10]: Also, I wonder why the "by-project-type" don't have a "name" ?

File: packages/eslint-plugin/lib/config/jsx-a11y.ts [hunk 16]
```diff
@@ -1,42 +1,36 @@
...
-                // There is a really good article that describes the issues with autoFocus and why it should be avoided:
-                // https://brucelawson.co.uk/2009/the-accessibility-of-html-5-autofocus/
-                // However, this issue is with screen readers and not with keyboard navigation.
-                // In Workleap, we use autoFocus in a lot of places to improve the user experience.
-                // Therefore, we are disabling this rule.
-                "jsx-a11y/no-autofocus": "off",
+const config: Linter.FlatConfig[] = await concat(
+    {
+        name: "Workleap/JsxA11y",
+        plugins: { "jsx-a11y": jsxA11yPlugin },
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/jsx-a11y.ts:9]: I am confused by this one, I thought plugins were not a thing anymore?

File: packages/eslint-plugin/eslint.config.js [hunk 17]
```diff
@@ -0,0 +1,10 @@
+import workleapPlugin from "@workleap/eslint-plugin";
+
+const config = [
+    ...workleapPlugin.configs.typescriptLibrary.map(conf => ({
```
Comment (patricklafrance) [inline packages/eslint-plugin/eslint.config.js:4]: Why does it requires the usage of a map?

File: packages/eslint-plugin/lib/config/jest.ts [hunk 18]
```diff
@@ -1,31 +1,20 @@
...
-                // Prefer spies to allow for automatic restoration
-                "jest/prefer-spy-on": "error",
-                // Gives better failure messages for array checks
-                "jest/prefer-to-contain": "error"
-            }
+const config: Linter.FlatConfig[] = await concat(
+    {
+        name: "Workleap/Jest",
+        files: [...testFiles, ...reactTestFiles],
+        ...jestPlugin.configs["flat/recommended"],
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/jest.ts:10]: Could something similar to `import packageJsonPluginRecommended from "eslint-plugin-package-json/configs/recommended";` also work for this one?

File: packages/eslint-plugin/lib/config/package-json.ts [hunk 19]
```diff
@@ -1,34 +1,31 @@
...
-                    ]
-                ],
-                // Doesn't support "workspace:*" at the moment.
-                "package-json/valid-package-def": "off",
-                // I am not sure why, this rule is triggering errors for valid paths.
-                "package-json/valid-repository-directory": "off"
-            }
+const config: Linter.FlatConfig[] = await concat(
+    {
+        ...(packageJsonPluginRecommended as Linter.FlatConfig),
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/package-json.ts:7]: Is the `as Linter.FlatConfig` still required?

File: packages/eslint-plugin/lib/config/package-json.ts [hunk 20]
```diff
@@ -1,34 +1,31 @@
...
+        name: "Workleap/Package.json",
+        files: ["package.json"],
+        rules: {
+            ...packageJsonPluginRecommended.rules,
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/package-json.ts:11]: Isn't the "concat" already taking care of merging the rules?

File: packages/eslint-plugin/lib/config/react.ts [hunk 21]
```diff
@@ -1,89 +1,93 @@
...
+            ],
+            "react/jsx-curly-spacing": ["warn", { children: true, when: "never" }]
 
-            }
         }
-    ]
-};
+    },
+    {
+        name: "Workleap/React",
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/react.ts:84]: Is it legit to have 2 objects with the same name?

File: packages/eslint-plugin/lib/config/by-project-type/monorepo-workspace.ts [hunk 22]
```diff
@@ -1,22 +1,28 @@
...
-        "plugin:@workleap/typescript",
-        "plugin:@workleap/jest",
-        "plugin:@workleap/testing-library",
-        "plugin:@workleap/mdx",
-        "plugin:@workleap/package-json",
-        "plugin:@workleap/yaml"
-    ],
-    rules: {
-        "package-json/valid-version": "off"
+const config: Linter.FlatConfig[] = await concat(
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/by-project-type/monorepo-workspace.ts:10]: I am wondering about the "await".

In the library documentation he is not awaiting the concat:

```js
import { concat } from 'eslint-flat-config-utils'

export default concat(
  {
    plugins: {},
    rules: {},
  },
  // It can also takes a array of configs:
  [
    {
      plugins: {},
      rules: {},
    }
    // ...
  ],
  // Or promises:
  Promise.resolve({
    files: ['*.ts'],
    rules: {},
  })
  // ...
)
```

File: packages/eslint-plugin/lib/config/react.ts (hunk 21 above)
Comment (patricklafrance) [reply packages/eslint-plugin/lib/config/react.ts:84]: What are the benefits of separating the react and hooks configs?

File: packages/eslint-plugin/lib/config/storybook.ts (hunk 5 above)
Comment (patricklafrance) [reply packages/eslint-plugin/lib/config/storybook.ts:22]: Well, if it's not necessary it kinds of gives the wrong impression to the next maintainer because he will initially believe that is it necessary, ideally, only what is required would be part of the config!

File: packages/eslint-plugin/lib/config/workleap.ts [hunk 23]
```diff
@@ -0,0 +1,18 @@
+import type { Linter } from "eslint";
+import StrictCSSModuleNames from "../rules/strict-css-modules-names.ts";
+
+const config: Linter.FlatConfig = {
+    plugins: {
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/workleap.ts:5]: If the idea if to import configs as JS objects or array and merge them on the consumer side, I don't get this plugin thing, what does it do?

File: packages/eslint-plugin/lib/config/typescript.ts [hunk 24]
```diff
@@ -1,76 +1,83 @@
...
-            plugins: ["@typescript-eslint"],
-            extends: [
-                "plugin:@typescript-eslint/eslint-recommended",
-                "plugin:@typescript-eslint/recommended"
-            ],
-            rules: {
-                // @typescript-eslint/recommended disables
-                "@typescript-eslint/no-non-null-assertion": "off",
+const config: Linter.FlatConfig[] = await concat(
+    js.configs.recommended,
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/typescript.ts:9]: I am not sure this is equivalent to:

```
extends: [
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended"
]
```

File: packages/eslint-plugin/lib/config/jsx-a11y.ts (hunk 16 above)
Comment (patricklafrance) [reply packages/eslint-plugin/lib/config/jsx-a11y.ts:9]: I feel like it either we define a plugin that takes care of merging everything or we merge everything ourself, not both?

File: packages/eslint-plugin/lib/config/mdx.ts [hunk 25]
```diff
@@ -1,16 +1,19 @@
...
+const config: Linter.FlatConfig[] = await concat(
+    {
+        name: "Workleap/MDX",
+        ...mdxPlugin.flat,
+        files: mdxFiles
+    },
+    {
+        name: "Workleap/MDX",
+        ...mdxPlugin.flatCodeBlocks,
+        files: mdxFiles
```
Comment (patricklafrance) [inline packages/eslint-plugin/lib/config/mdx.ts:15]: Is there a reason why it requires to be define in 2 blocks?

Also, are we still getting only the "recommended" rules?

**Comment by patricklafrance** [comment]
There's an unresolved links in the migration docs page:

```
  INPUT: C:\Dev\gsoft\wl-web-configs\docs
 OUTPUT: [in-memory]
WARNING: [eslint\flat-config-migration.md:176] URL "wl-web-configs/eslint/advanced-composition/" was not resolved.
         Please use --verbose for more details

```

File: docs/eslint/flat-config-migration.md [hunk 26]
```diff
@@ -0,0 +1,206 @@
...
+
+We recommend using one of the "by project type" configurations for simplicity and consistency. But if you need to customize further, you can choose to combine [any individual configs](wl-web-configs/eslint/advanced-composition/). Here, we'll compose a config for a project that uses React and TypeScript. 
+
+```javascript
+import workleapPlugin from "@workleap/eslint-config";
+
+const config = [
+    ...workleapPlugin.configs.core,
+    ...workleapPlugin.configs.typescript,
+    ...workleadPlugin.configs.react
```
Comment (patricklafrance) [inline docs/eslint/flat-config-migration.md:184]: Should we recommend using antfu "concat" function here?

File: docs/eslint/flat-config-migration.md [hunk 27]
```diff
@@ -0,0 +1,206 @@
...
+```
+
+Some rules may need to be overridden within each nested config object. Since flat configs are entirely JavaScript, we can manipulate the underlying configuration objects directly. For example, to change the file type used by a config object: 
+```javascript
+import workleapPlugin from "@workleap/eslint-config";
+
+const config = [
+    ...workleapPlugin.configs.core,
+    ...workleapPlugin.configs.typescript,
+    ...workleadPlugin.configs.react.map(conf => (
```
Comment (patricklafrance) [inline docs/eslint/flat-config-migration.md:197]: Should we recommend using antfu "concat" function here?

File: packages/eslint-plugin/lib/config/react.ts (hunk 21 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/react.ts:84]: The double name is legit, the rules will both show up as being part of the React config. There's no functional up or down side to keeping them separate. Just makes it easier to see which is which.

File: docs/eslint/flat-config-migration.md [hunk 28]
```diff
@@ -0,0 +1,206 @@
...
+
+## High level differences
+
+Previously, ESLint allowed you to use multiple formats to define your config files. Flat config can only use JavaScript. You can import plugins and pre-made configuration objects directly, and manipulate them as necessary.
+
+The `extends` keyword has been removed. Now you simply add multiple _configuration objects_ to an array. Configuration objects will cascade, similar to the `overrides` block of the old config.
+
+The `.eslintignore` file is no longer valid. If you need to exclude files from linting, add them to a configuration block under the `ignores` key.
+
+Config files no longer rely on custom resolution implemented by ESLint. Since they import modules directly, they now rely on Node file resolution. Config files no longer merge down the file tree, either. Every config file acts as if it is the root config file. This means if you have nested configs as well as a true root config (like in a monorepo), you will want to set your top-level config to ignore directories that have their own config files.
```
[the remark this answers is not in the document]
Comment (bsokol-wl) [reply docs/eslint/flat-config-migration.md:20]: No, I need to reword this.

File: docs/eslint/flat-config-migration.md [hunk 29]
```diff
@@ -0,0 +1,206 @@
...
+
+## Basic migration steps
+
+1. Create a file called `eslint.config.js`. `@workleap/eslint-config` is published in ESM, so if your project `type` in `package.json` is not `module`, then you should create an `eslint.config.mjs` file instead.
```
[the remark this answers is not in the document]
Comment (bsokol-wl) [reply docs/eslint/flat-config-migration.md:24]: Yep

File: docs/eslint/flat-config-migration.md [hunk 30]
```diff
@@ -0,0 +1,206 @@
...
+        ignores: ["node_modules/", "dist/"]
+    }
+];
+
+export default config;
+```
+
+## Recommended setup - By project type
+
+`@workleap/eslint-config` exposes some pre-built configs based on common project types. Each of these configs are properly set up for **JavaScript**, **TypeScript**, **Jest**, **Testing Library**, **MDX**, **package.json**, and **YAML**.
```
[the remark this answers is not in the document]
Comment (bsokol-wl) [reply docs/eslint/flat-config-migration.md:57]: Right

File: docs/eslint/flat-config-migration.md [hunk 31]
```diff
@@ -0,0 +1,206 @@
...
+import workleapPlugin from "@workleap/eslint-config";
+
+const config = [
+    {
+        ignores: ["node_modules/", "dist/"]
+    },
+    ...workleapPlugin.configs.webApplication,
+    {
+        rules: {
+            'react/jsx-uses-vars': 'error',
```
[the remark this answers is not in the document]
Comment (bsokol-wl) [reply docs/eslint/flat-config-migration.md:93]: Got it

File: docs/eslint/flat-config-migration.md [hunk 32]
```diff
@@ -0,0 +1,206 @@
...
+];
+
+export default config;
+```
+
+### Top level
+
+Create a new `eslint.config.js` at the root of your project. Import the monorepo workspace config.
+
+In order to make it easier to scope config files to monorepo packages, we recommend using [eslint-flat-config-utils](https://github.com/antfu/eslint-flat-config-utils). The `concat` function makes it easier to join multiple configs together. Wrap your configuration objects in the `concat` function because it will automatically merege arrays.
```
[the remark this answers is not in the document]
Comment (bsokol-wl) [reply docs/eslint/flat-config-migration.md:136]: Fixed

File: package.json [hunk 33]
```diff
@@ -38,10 +38,10 @@
     "devDependencies": {
         "@changesets/changelog-github": "0.5.0",
         "@changesets/cli": "2.27.1",
-        "@typescript-eslint/parser": "7.6.0",
         "@workleap/eslint-plugin": "workspace:*",
         "@workleap/typescript-configs": "workspace:*",
         "eslint": "8.57.0",
+        "eslint-flat-config-utils": "^0.2.4",
```
[the remark this answers is not in the document]
Comment (bsokol-wl) [reply package.json:44]: Removed

File: packages/eslint-plugin/package.json [hunk 34]
```diff
@@ -35,17 +36,13 @@
         "build": "tsup"
     },
     "peerDependencies": {
-        "@typescript-eslint/parser": "*",
-        "eslint": "*",
+        "eslint": "8.57.0",
         "typescript": "*"
     },
     "peerDependenciesMeta": {
         "@typescript-eslint/parser": {
```
[the remark this answers is not in the document]
Comment (bsokol-wl) [reply packages/eslint-plugin/package.json:43]: removed

File: packages/eslint-plugin/package.json [hunk 35]
```diff
@@ -55,6 +52,7 @@
         "@swc/helpers": "0.5.9",
         "@swc/jest": "0.2.36",
         "@types/eslint": "8.56.9",
+        "@types/eslint__js": "^8.42.3",
```
[the remark this answers is not in the document]
Comment (bsokol-wl) [reply packages/eslint-plugin/package.json:55]: removed

File: packages/browserslist-config/eslint.config.mjs [hunk 36]
```diff
@@ -0,0 +1,5 @@
+import workleapPlugin from "@workleap/eslint-plugin";
+
+const config = workleapPlugin.configs.typescriptLibrary;
```
[the remark this answers is not in the document]
Comment (bsokol-wl) [reply packages/browserslist-config/eslint.config.mjs:3]: Yeah, if they mutate it directly. I can do an object spread like in the other ones. Probably better for consistency anyway.

File: packages/eslint-plugin/lib/config/core.ts (hunk 14 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/core.ts:8]: Yes, because the type of `js.configs.recommended` is actually a subset of `Linter.FlatConfig` (it only contains the rules) and it confuses TypeScript as to what exactly should be in the array. I makes it mark everything that isn't `rules` in the second object as unknown keys. Since it's a subset, then casting it to `Linter.FlatConfig` clears everything up.

File: packages/eslint-plugin/lib/config/core.ts (hunk 15 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/core.ts:10]: The slash is the recommended name, and this is only used in debug view, and also in the inspector. The by project type ones don't have their own custom rules, so there is nothing to name. I had at one point toyed around with a function that did a deep rewrite of the names so that it appended the by project name to everything, but it seemed like a lot of work for not a lot of benefit.

File: packages/eslint-plugin/lib/config/jsx-a11y.ts (hunk 16 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/jsx-a11y.ts:9]: At this point, plugins are just used to define custom rule behavior. The one Workleap custom rule requires the `workleapPlugin` to be provided as a plugin. If anyone wanted to just use that rule, they would have to provide to provide the `workleapPlugin` as a plugin.

File: packages/eslint-plugin/eslint.config.js (hunk 17 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/eslint.config.js:4]: because the typescriptLibrary config is an array of config objects, and I want the ignore to apply to the objects within this array, and not to all configs. But as I'm thinking through it, the top level `extends` will scope it down, so I'll change this to not map.

File: packages/eslint-plugin/lib/config/jest.ts (hunk 18 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/jest.ts:10]: No, "flat/recommended" is actually an object key, and not a file path, so it doesn't know how to import that. The package.json one is a true file path, and it doesn't actually have a default export at the root. There's no consistency between these libraries at the moment, even the one ones that have updated.

File: packages/eslint-plugin/lib/config/package-json.ts (hunk 19 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/package-json.ts:7]: Yes, for similar reasons as the ESlint casting I explained in a different comment.

File: packages/eslint-plugin/lib/config/package-json.ts (hunk 20 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/package-json.ts:11]: No, concat just merges and flattens arrays, it doesn't do a deep merge. It still produces an array of config objects.

File: packages/eslint-plugin/lib/config/by-project-type/monorepo-workspace.ts (hunk 22 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/by-project-type/monorepo-workspace.ts:10]: `concat` always returns a promise, even if none of the things passed in are promises. So the issue I was struggling with was when exactly to handle this promise. ESLint can actually take a promise that resolves to a config array (so it can take the direct output of `concat`), but it can't take an array of promises.

If I don't await at the top of each config module, then each one would export a promise. This forces any consumer to use either `concat` or `Promise.All` + some array spread to resolve all of these before it goes to ESLint. And it gets a little more difficult if you need to overwrite values in your config, because you'd need to resolve the promise first and then map or reduce that into the config you want. That all adds just a little more complexity for setup than a typical ESLint config would require.

Top level await there seemed like the best solution.

File: packages/eslint-plugin/lib/config/storybook.ts (hunk 5 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/storybook.ts:22]: The thing I'm not clear on is if it needs to be provided twice. I think it does since the file patterns are different, so ESLint needs to know these rules apply to both file globs. At worst it just overwrites itself. None of this would be necessary once the Storybook plugin updates for flat config. I have to do this because it's not in the right format. Ideally, this would just be part of the Storybook plugin's recommended config and you wouldn't have to manually provide it. So I agree with you there, this shouldn't be necessary, but for now, it is.

File: packages/eslint-plugin/lib/config/typescript.ts (hunk 24 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/typescript.ts:9]: This is how they recommend doing it now in the typescript-eslint docs. Instead of re-providing the ESLint rules, they just have you import them from ESLint directly.

File: packages/eslint-plugin/lib/config/workleap.ts (hunk 23 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/workleap.ts:5]: In ESLint plugins that have been properly updated, this `plugins` key is just part of the recommended config, and you don't have to manually provide it. Ideally, once everything is updated, you don't need to manually provide it unless you need to override something.

The other reason to provide it is to change the prefix of the rules. For example, typescript-eslint recommends manually providing the plugin to the "@typescript-eslint" key, so that any rules that start with `@typescript-eslint/` will be matched with those rules. typescript-eslint changed their package name so it isn't but they didn't want everyone to have to change all the rules, so they recommend this for backwards compatibility.

This is the same reason I provide the plugin directly for the react-hooks plugin, even though the rest of the plugin format is already technically compatible with flat config.

Long story short:

1. The purpose of the `plugin` key is to provide rules definitions to ESLint
2. Ideally, this would be part of the recommended config and you wouldn't have to manually provide it
3. There are still legitimate reasons to provide it manually, usually if the plugin's docs tell you to

File: packages/eslint-plugin/lib/config/mdx.ts (hunk 25 above)
Comment (bsokol-wl) [reply packages/eslint-plugin/lib/config/mdx.ts:15]: I went back to the docs on this one and realized that code blocks is part of flat config, so there is no need to do it twice. I do need to add a processor to hit the code blocks. The recommended rules should be part of the flat config.

File: docs/eslint/flat-config-migration.md (hunk 26 above)
Comment (bsokol-wl) [reply docs/eslint/flat-config-migration.md:184]: Yeah, I'll update this

File: packages/eslint-plugin/lib/config/core.ts (hunk 15 above)
Comment (patricklafrance) [reply packages/eslint-plugin/lib/config/core.ts:10]: Got it thanks!

Could you name like: "workleap/core" thought? Feels more JS and less .NET hehe

File: packages/eslint-plugin/package.json [hunk 37]
```diff
@@ -63,23 +61,22 @@
...
-        "eslint-plugin-react-hooks": "^4.6.0",
+        "eslint-plugin-react-hooks": "^4.6.2",
         "eslint-plugin-storybook": "^0.8.0",
-        "eslint-plugin-testing-library": "^6.2.1",
-        "eslint-plugin-yml": "^1.14.0",
+        "eslint-plugin-testing-library": "^6.2.2",
+        "globals": "^15.1.0",
         "jsonc-eslint-parser": "^2.4.0",
-        "yaml-eslint-parser": "^1.2.2"
+        "typescript-eslint": "^7.8.0"
```
[the remark this answers is not in the document]
Comment (alexasselin008) [reply packages/eslint-plugin/package.json:76]: i can't remember. The inspiration project i looked at was shopify's and they have it in dependencies, not peer dependencies. Most likely we changed in because we had issues with versionning in our projects
